flink-runtime/src/main/java/org/apache/flink/runtime/checkpoint/CheckpointCoordinator.java

```diff
@@ -485,76 +481,151 @@ public boolean isShutdown() {
-						job);
-				throw new CheckpointException(CheckpointFailureReason.NOT_ALL_REQUIRED_TASKS_RUNNING);
-			} else if (ee.getState() == ExecutionState.RUNNING) {
-				executions[i] = ee;
-			} else {
-				LOG.info("Checkpoint triggering task {} of job {} is not in state {} but {} instead. Aborting checkpoint.",
-						tasksToTrigger[i].getTaskNameWithSubtaskIndex(),
-						job,
-						ExecutionState.RUNNING,
-						ee.getState());
-				throw new CheckpointException(CheckpointFailureReason.NOT_ALL_REQUIRED_TASKS_RUNNING);
+			if (isTriggering || !triggerRequestQueue.isEmpty()) {
+				// we can't trigger checkpoint directly if there is a trigger request being processed
+				// or queued
+				triggerRequestQueue.add(new CheckpointTriggerRequest(
+					timestamp,
+					props,
+					externalSavepointLocation,
+					isPeriodic,
+					advanceToEndOfTime,
+					onCompletionPromise));
+				return onCompletionPromise;
 			}
 		}
+		startTriggeringCheckpoint(
+			timestamp,
+			props,
+			externalSavepointLocation,
+			isPeriodic,
+			advanceToEndOfTime,
+			onCompletionPromise);
+		return onCompletionPromise;
+	}
 
-		// next, check if all tasks that need to acknowledge the checkpoint are running.
-		// if not, abort the checkpoint
-		Map<ExecutionAttemptID, ExecutionVertex> ackTasks = new HashMap<>(tasksToWaitFor.length);
+	private void startTriggeringCheckpoint(
+		long timestamp,
+		CheckpointProperties props,
+		@Nullable String externalSavepointLocation,
+		boolean isPeriodic,
+		boolean advanceToEndOfTime,
+		CompletableFuture<CompletedCheckpoint> onCompletionPromise) {
 
-		for (ExecutionVertex ev : tasksToWaitFor) {
-			Execution ee = ev.getCurrentExecutionAttempt();
-			if (ee != null) {
-				ackTasks.put(ee.getAttemptId(), ev);
-			} else {
-				LOG.info("Checkpoint acknowledging task {} of job {} is not being executed at the moment. Aborting checkpoint.",
-						ev.getTaskNameWithSubtaskIndex(),
-						job);
-				throw new CheckpointException(CheckpointFailureReason.NOT_ALL_REQUIRED_TASKS_RUNNING);
+		try {
+			// make some eager pre-checks
+			synchronized (lock) {
+				preCheckBeforeTriggeringCheckpoint(isPeriodic, props.forceCheckpoint());
 			}
-		}
 
-		// we will actually trigger this checkpoint!
+			final Execution[] executions = getTriggerExecutions();
+			final Map<ExecutionAttemptID, ExecutionVertex> ackTasks = getAckTasks();
+
+			// we will actually trigger this checkpoint!
+			Preconditions.checkState(!isTriggering);
+			isTriggering = true;
+
+			final CompletableFuture<PendingCheckpoint> pendingCheckpointCompletableFuture =
+				initializeCheckpoint(props, externalSavepointLocation)
+					.thenApplyAsync((checkpointIdAndStorageLocation) -> createPendingCheckpoint(
+						timestamp,
+						props,
+						ackTasks,
+						isPeriodic,
+						checkpointIdAndStorageLocation.checkpointId,
+						checkpointIdAndStorageLocation.checkpointStorageLocation,
+						onCompletionPromise), timer);
```

Comment:
nit: `timer` in new line? It's not easy to deduce where the `timer` parameter belongs to. Maybe:
```
	.thenApplyAsync(
		(checkpointIdAndStorageLocation) -> createPendingCheckpoint(
			timestamp,
			props,
			ackTasks,
			isPeriodic,
			checkpointIdAndStorageLocation.checkpointId,
			checkpointIdAndStorageLocation.checkpointStorageLocation,
			onCompletionPromise),
		timer);
```
and in other places (I think everywhere where you are specifying executor)?